core/src/display_object.rs

```diff
@@ -1331,3 +1364,71 @@ bitflags! {
         const LOCK_ROOT                = 1 << 6;
     }
 }
+
+/// Represents the sound transfomr of sounds played inside a Flash MovieClip.
+/// In AVM1, this is returned by `Sound.getTransform`.
+/// In AVM2, this is returned by `Sprite.soundTransform`.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct SoundTransform {
+    pub volume: i32,
+    pub left_to_left: i32,
+    pub left_to_right: i32,
+    pub right_to_left: i32,
+    pub right_to_right: i32,
+}
+
+impl SoundTransform {
+    pub const MAX_VOLUME: i32 = 100;
+
+    pub fn concat(&mut self, other: &SoundTransform) {
+        const MAX_VOLUME: i64 = SoundTransform::MAX_VOLUME as i64;
+        self.volume = (i64::from(self.volume) * i64::from(other.volume) / MAX_VOLUME) as i32;
+
+        let ll0 = i64::from(self.left_to_left);
+        let lr0 = i64::from(self.left_to_right);
+        let rl0 = i64::from(self.right_to_left);
+        let rr0 = i64::from(self.right_to_right);
+        let ll1 = i64::from(other.left_to_left);
+        let lr1 = i64::from(other.left_to_right);
+        let rl1 = i64::from(other.right_to_left);
+        let rr1 = i64::from(other.right_to_right);
+        self.left_to_left = ((ll0 * ll1 + rl0 * lr1) / MAX_VOLUME) as i32;
+        self.left_to_right = ((lr0 * ll1 + rr0 * lr1) / MAX_VOLUME) as i32;
+        self.right_to_left = ((ll0 * rl1 + rl0 * rr1) / MAX_VOLUME) as i32;
+        self.right_to_right = ((lr0 * rl1 + rr0 * rr1) / MAX_VOLUME) as i32;
+    }
+
+    /// Sets this transform the pan from -1.0 left to 1.0 right.
+    pub fn pan(&self) -> i32 {
+        if self.left_to_left != Self::MAX_VOLUME {
+            Self::MAX_VOLUME - self.left_to_left.abs()
+        } else {
+            self.right_to_right.abs() - Self::MAX_VOLUME
+        }
+    }
+
+    /// Sets this transform the pan from -1.0 left to 1.0 right.
+    pub fn set_pan(&mut self, pan: i32) {
+        if pan >= 0 {
+            self.left_to_left = Self::MAX_VOLUME - pan;
+            self.right_to_right = Self::MAX_VOLUME;
+        } else {
+            self.left_to_left = Self::MAX_VOLUME;
+            self.right_to_right = Self::MAX_VOLUME - pan;
+        }
+        self.left_to_right = 0;
+        self.right_to_left = 0;
+    }
+}
+
+impl Default for SoundTransform {
+    fn default() -> Self {
+        Self {
+            volume: 100,
+            left_to_left: 100,
+            left_to_right: 0,
+            right_to_left: 0,
+            right_to_right: 100,
```

Comment:
I don't follow; this is a SoundTransform for a display object, not related to a specific sound, so it's always "stereo". Could you link me to the text in the (SWF? AS2?) spec you are reading?

Reply:
Ok, I see the paragraph in the AS2 spec. It's a little misleading because you don't get the transform for a specific sound, so there's no way to "see" the transform the docs are describing -- I think the text is more for descriptive purposes about how a mono sound gets upmixed to stereo. Even in AS3 where you can read a SoundTransform directly from a SoundChannel, it returns `1, 0, 0, 1` for a mono sound. This is the correct default for a SoundTransform because it's an identity transform.